godel_surface_detection/src/scan/profilimeter_scan.cpp

```diff
@@ -0,0 +1,237 @@
+#include "godel_surface_detection/scan/profilimeter_scan.h"
+#include <ros/ros.h>
+using namespace godel_process_path;
+
+// Local types
+namespace
+{
+  // stub for gift-wrapping algorithm to compute convex hull
+  // stub for rotating-calipers algorithm to compute min rectangle
+
+  /**
+   * Rotated rectangle structure for use with striping algorithm
+   * (x,y) is the center (of mass)
+   * w, h are the edge lengths
+   * a is the clockwise rotation from upright
+   */
+  struct RotatedRect
+  {
+    double x, y, w, h, a;
+  };
+
+  RotatedRect simpleBoundingBox(const PolygonBoundary& boundary)
+  {
+    double min_x = boundary[0].x;
+    double min_y = boundary[0].y;
+    double max_x = boundary[0].x;
+    double max_y = boundary[0].y;
+
+    for (std::size_t i = 1; i < boundary.size(); ++i)
+    {
+      const PolygonPt& pt = boundary[i];
+      min_x = std::min(pt.x, min_x);
+      min_y = std::min(pt.y, min_y);
+      max_x = std::max(pt.x, max_x);
+      max_y = std::max(pt.y, max_y); 
+    }
+
+    RotatedRect result;
+    result.a = 0.0; // non-rotated, hence the simple bounding box
+    result.x = (max_x + min_x) / 2;
+    result.y = (max_y + min_y) / 2;
+    result.w = max_x - min_x;
+    result.h = max_y - min_y;
+    return result;
+  }
+
+  // Vertically or horizontally?
+  std::vector<RotatedRect> sliceBoundingBox(const RotatedRect& bbox, double slice_width, double overlap)
+  {
+    // we assume slice_width > overlap
+    int n_slices;
+    float dx, dy; // unit vector
+    float x, y; // start position
+    float w, h;
+
+    float adjusted_width = slice_width - overlap;
+
+    // Determine which way the slice should go
+    if (bbox.w < bbox.h)
+    {
+      // walk horizontally, vertical slices
+      n_slices = static_cast<int>(bbox.w / (adjusted_width)) + 1;
+      
+      dx = std::cos(bbox.a);
+      dy = std::sin(bbox.a);
+
+      x = bbox.x - dx * (bbox.w - adjusted_width) / 2.0;
+      y = bbox.y - dy * (bbox.w - adjusted_width) / 2.0;
+
+      w = adjusted_width;
+      h = bbox.h;
+    }
+    else
+    {
+      // walk vertically, horizontal slices
+      n_slices = static_cast<int>(bbox.h / (adjusted_width)) + 1;
+      
+      dx = -std::sin(bbox.a);
+      dy = std::cos(bbox.a);
+
+      x = bbox.x - dx * (bbox.h - adjusted_width) / 2.0;
+      y = bbox.y - dy * (bbox.h - adjusted_width) / 2.0; 
+
+      w = bbox.w;
+      h = adjusted_width;
+    }
+    
+    std::vector<RotatedRect> slices;
+    slices.reserve(n_slices);
+
+    dx *= adjusted_width;
+    dy *= adjusted_width;
+
+    for (int i = 0; i < n_slices; ++i)
+    {
+      RotatedRect temp;
+      temp.a = bbox.a;
+      temp.x = x;
+      temp.y = y;
+      temp.w = w;
+      temp.h = h;
+      slices.push_back(temp);
+      x += dx;
```

Comment:
Don't know if this is relevant but the last update to **x** and **y** don't get saved in the **temp** object